src/shogun/machine/gp/FITCInferenceBase.h

```diff
@@ -0,0 +1,291 @@
+
+	/** get the noise for inducing points
+	 *
+	 * @return noise noise for inducing points
+	 *
+	 */
+	virtual float64_t get_inducing_noise();
+
+	/** returns derivative of negative log marginal likelihood wrt latent features (input)
+	 * Note that in order to call this method, kernel must support FITC inference
+	 *
+	 * @return derivative of negative log marginal likelihood
+	 */
+	virtual SGVector<float64_t> get_derivative_wrt_latent_features(const TParameter* param)=0;
+
+	/** returns mean vector \f$\mu\f$ of the Gaussian distribution
+	 * \f$\mathcal{N}(\mu,\Sigma)\f$, which is an approximation to the
+	 * posterior:
+	 *
+	 * \f[
+	 * p(f|y) \approx q(f|y) = \mathcal{N}(\mu,\Sigma)
+	 * \f]
+	 *
+	 * in case if particular inference method doesn't compute posterior
+	 * \f$p(f|y)\f$ exactly, and it returns covariance matrix \f$\Sigma\f$ of
+	 * the posterior Gaussian distribution \f$\mathcal{N}(\mu,\Sigma)\f$
+	 * otherwise.
+	 *
+	 * @return mean vector
+	 */
+	virtual SGVector<float64_t> get_posterior_mean();
+
+	/** returns covariance matrix \f$\Sigma\f$ of the Gaussian distribution
+	 * \f$\mathcal{N}(\mu,\Sigma)\f$, which is an approximation to the
+	 * posterior:
+	 *
+	 * \f[
+	 * p(f|y) \approx q(f|y) = \mathcal{N}(\mu,\Sigma)
+	 * \f]
+	 *
+	 * in case if particular inference method doesn't compute posterior
+	 * \f$p(f|y)\f$ exactly, and it returns covariance matrix \f$\Sigma\f$ of
+	 * the posterior Gaussian distribution \f$\mathcal{N}(\mu,\Sigma)\f$
+	 * otherwise.
+	 *
+	 * @return covariance matrix
+	 */
+	virtual SGMatrix<float64_t> get_posterior_covariance();
+
+protected:
+	/** convert latent features and features to the same represention
+	 *
+	 * Note that these two kinds of features can different types.
+	 * The reasons are listed below.
+	 * 1. The type of the gradient wrt latent features is float64_t, which is used to update latent features
+	 * 2. Reason 1 implies that the type of latent features can be float64_t while the type of features does not required
+	 * as float64_t 
+	 * 3. Reason 2 implies that the type of features must be a subclass of CDotFeatures, which can represent features as
+	 * float64_t
+	 */
+	virtual void convert_features();
+
+	/** check whether features and latent features are set
+	 */
+	virtual void check_features();
+
+	/** check if members of object are valid for inference */
+	virtual void check_members() const;
+
+	/** update train kernel matrix */
+	virtual void update_train_kernel();
+
+	/** returns derivative of negative log marginal likelihood wrt parameter of
+	 * CInferenceMethod class
+	 *
+	 * @param param parameter of CInferenceMethod class
+	 *
+	 * @return derivative of negative log marginal likelihood
+	 */
+	virtual SGVector<float64_t> get_derivative_wrt_inference_method(
+			const TParameter* param)=0;
+
+	/** returns derivative of negative log marginal likelihood wrt parameter of
+	 * likelihood model
+	 *
+	 * @param param parameter of given likelihood model
```

Comment:
state math of log marginal likelihood?

Also maybe say that it is the *approximate* log marginal likelihood